Add Librato provider

henrikhodne opened the pull request:
This provider mainly adds support for Librato Spaces and charts in these spaces for now.

henrikhodne commented on builtin/providers/librato/resource_librato_space_chart.go:
```diff
@@ -0,0 +1,408 @@
+				stream.UnitsLong = librato.String(v)
+			}
+			// TODO: This isn't correct, since it means you can't set 0 as the minimum
+			if v, ok := streamData["min"].(float64); ok && v != 0 {
```
I'm not sure how to best discover the difference between a zero value and the value not being set for nested blocks. This TODO message occurs a few other places too, but they are all essentially the same issue.

henrikhodne commented:
I forgot to push the latest commits of the go-librato library, I believe all the test failures are related to that (I pushed the changes now).

henrikhodne commented:
I have tested the following by making actual Terraform files in a different repository:

- Creating Spaces
- Renaming a space
- Adding charts to a space that isn't created yet (creating spaces and charts in one `terraform apply`)
- Adding charts to an existing space
- Deleting charts
- Deleting spaces
- Editing some attributes in a chart

BRMatt commented:
Wow this is a great idea! :heart:

obfuscurity commented:
Very cool. Much :heart::heart::heart:.

obfuscurity commented:
@jen20 According to your comment [here](https://github.com/hashicorp/terraform/issues/3983#issuecomment-158119130), the plan was to get this merged in post-0.6.7. Looks like 0.6.8 was rolled out a month ago, any immediate plans (or obstacles) for getting this merged?

/cc @carl-youngblood @ralph-tice

henrikhodne commented:
Sorry for letting this become stale. I see there are some other implementation of this too that adds support for more Librato APIs. I can work on getting some of that merged into my API client and this PR too this week.

obfuscurity commented:
:heart:

jen20 commented:
Hi! This is definitely something I would like to get merged - one of the things we've been trying to do is get the [acceptance tests running in Travis](https://travis-ci.org/hashicorp/terraform/builds) for new providers we are merging, so I'll need to set up a test account with Librato for that. Without looking I'd guess the free tier should be sufficient for that.

@henrikhodne - sorry this has taken so long to review. Is there any chance you could rebase this to master, squash to one commit and force push back onto the branch this PR is tracking? Thanks!

henrikhodne commented:
No problem, I'm sorry I didn't get around to implement more things. I tried to squash everything together, but for some reason c10df6afea4e8ff483a8a538683e844e70911dbb got included as a commit here? Not quite sure how that happened, trying to invoke some Git-fu to see if I can fix that.

henrikhodne commented:
There we go, squashed and rebased! (Was very confused why the commit log locally on my machine and on GitHub was completely different when Git claimed there was no new commits to push, but one more force push seemed to fix it).

henrikhodne commented:
Looks like the build failed, and I'm getting some errors locally due to some changes made in the API client. I'll try to get that fixed later this evening (CET).

ralph-tice commented:
has "We'll check to ensure that the same range of features are supported." been done, re: #3983 ?

phinze commented:
Hey there! Just attempted a run through of the acceptance tests and got a couple of 404s.

```
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccLibratoSpaceChart_Basic
--- PASS: TestAccLibratoSpaceChart_Basic (2.69s)
=== RUN   TestAccLibratoSpaceChart_Full
--- FAIL: TestAccLibratoSpaceChart_Full (0.28s)
        testing.go:148: Step 0 error: Error applying: 1 error(s) occurred:

                * librato_space_chart.foobar: Error creating Librato space chart Foo Bar: POST https://metrics-api.librato.com/v1/spaces/157389/charts: 404  Request errors: Not Found,.
=== RUN   TestAccLibratoSpaceChart_Updated
--- FAIL: TestAccLibratoSpaceChart_Updated (0.18s)
        testing.go:148: Step 0 error: Error applying: 1 error(s) occurred:

                * librato_space_chart.foobar: Error creating Librato space chart Foo Bar: POST https://metrics-api.librato.com/v1/spaces/157390/charts: 404  Request errors: Not Found,.
=== RUN   TestAccLibratoSpace_Basic
--- FAIL: TestAccLibratoSpace_Basic (0.16s)
        testing.go:148: Step 0 error: Check failed: Check 1/3 error: GET https://metrics-api.librato.com/v1/spaces/157391: 404  Request errors: Not Found,.
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/builtin/providers/librato   3.332s
```

Any idea what might be happening to cause these?

phinze commented:
Also it seems like the tests might be leaking Spaces based on my account after running them:

<img width="634" alt="spaces___librato" src="https://cloud.githubusercontent.com/assets/37534/13832232/31125596-eba8-11e5-976f-3f77c7ec23a7.png">

henrikhodne commented:
Hmm, I can reproduce this locally too now. Checking to see what's up.

henrikhodne commented:
Interestingly, when I run the tests one at a time, no resources are leaked.

I also noticed that when deleting something through the Librato API, you can sometimes still retrieve the resource for a short while afterwards, which I think is causing some of the errors.

henrikhodne commented:
Looks like the Librato API sometimes being delayed on update is causing the leaked resources too. What it looks like is happening:

1. Terraform creates space
2. Terraform tries to create space chart, but Librato API says the space doesn't exist
3. Terraform exits due to error
4. Librato still has the space that was created

I emailed Librato to see if there's a way to work around this, or if I just need to add some sleeps in the code to make sure that all the servers know about the resource before continuing (and if so, how long one should sleep for).

phinze commented:
Thanks for looking into this, @henrikhodne - in other cases of eventually consistent APIs we end up polling a `Get` operation for a period of time after a `Create` until it shows up correctly.

So in this case, after the [Create API call](https://github.com/hashicorp/terraform/pull/3371/files#diff-3189ce3f6f360a8720a62ba53f1fa5e0R38) we'd do something like:

```go
resource.Retry(1*time.Minute, func() *resource.RetryError {
	_, _, err := client.Spaces.Get(uint(id))
	if err != nil {
		if errResp, ok := err.(*librato.ErrorResponse); ok && errResp.Response.StatusCode == 404 {
			return resource.RetryableError(err)
		}
		return resource.NonRetryableError(err)
	}
        return nil
})
```

Does that make sense?

henrikhodne commented:
@phinze Thanks, I was meaning to ask if you had a pattern for this, since I imagine this is not the first eventually consistent API you've had to deal with ;)

The latest commit has that pattern implemented for creates and deletes. I'm not sure if they are needed for updates?

The tests pass, at least:

```
➜  terraform git:(librato-provider) ✗ make testacc TEST=./builtin/providers/librato
==> Checking that code complies with gofmt requirements...
/home/henrik/gopath/bin/stringer
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/librato -v  -timeout 120m
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccLibratoSpaceChart_Basic
--- PASS: TestAccLibratoSpaceChart_Basic (8.93s)
=== RUN   TestAccLibratoSpaceChart_Full
--- PASS: TestAccLibratoSpaceChart_Full (9.59s)
=== RUN   TestAccLibratoSpaceChart_Updated
--- PASS: TestAccLibratoSpaceChart_Updated (11.98s)
=== RUN   TestAccLibratoSpace_Basic
--- PASS: TestAccLibratoSpace_Basic (5.47s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/librato	35.975s
```

henrikhodne commented:
The question I asked in https://github.com/hashicorp/terraform/pull/3371#discussion_r40859707 still stands though. Is there a way to detect a set `0` and something being `0` because it's not set? I seem to recall there being a change where `-1` was being used by some resources for "not set", do you have any examples of that I can look at?

phinze commented:
> Is there a way to detect a set 0 and something being 0 because it's not set?

Heh, we're actually discussing this just now over in #5694. Due to some core limitations, we don't have the ability to directly detect from a provider if a value is set or unset in config.

You're correct that the workaround we've used in other places is to use a "sentinel value" as a `Default` so that within the provider code you can detect it and treat it as "unset".

henrikhodne commented:
Ah, I see how it's done. Made `NaN` a sentinel in this case, since the API doesn't accept that as a valid value.

josephruscio commented:
@henrikhodne will get someone at Librato to clarify the expectations on the Spaces API here cc @nextmat

phinze commented:
This is close! Getting the following test failure on the latest version:

```
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccLibratoSpaceChart_Basic
--- PASS: TestAccLibratoSpaceChart_Basic (5.37s)
=== RUN   TestAccLibratoSpaceChart_Full
--- FAIL: TestAccLibratoSpaceChart_Full (4.12s)
        testing.go:148: Step 0 error: Error applying: 1 error(s) occurred:

                * librato_space_chart.foobar: Error creating Librato space chart Foo Bar: POST https://metrics-api.librato.com/v1/spaces/158674/charts: 400  Parameter errors: streams: no metric 'cpu' exists,.
=== RUN   TestAccLibratoSpaceChart_Updated
--- PASS: TestAccLibratoSpaceChart_Updated (6.17s)
=== RUN   TestAccLibratoSpace_Basic
--- PASS: TestAccLibratoSpace_Basic (3.38s)
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/builtin/providers/librato        19.059s
```

henrikhodne commented:
I'm on vacation through the end of March, so I probably won't be able to look into this until then.

To get around it, what you can do is sign in to your Librato account and there should be an option somewhere to generate some sample data.

I'd much rather have this be fixed in the tests, though. What I think I'll do is merge the PR on the Librato client library, and then change the tests to create the metrics it uses before the tests.

> On Mar 20, 2016, at 12:43, Paul Hinze <notifications@github.com> wrote:
> 
> This is close! Getting the following test failure on the latest version:
> 
> === RUN   TestProvider
> --- PASS: TestProvider (0.00s)
> === RUN   TestProvider_impl
> --- PASS: TestProvider_impl (0.00s)
> === RUN   TestAccLibratoSpaceChart_Basic
> --- PASS: TestAccLibratoSpaceChart_Basic (5.37s)
> === RUN   TestAccLibratoSpaceChart_Full
> --- FAIL: TestAccLibratoSpaceChart_Full (4.12s)
>         testing.go:148: Step 0 error: Error applying: 1 error(s) occurred:
> 
>                 * librato_space_chart.foobar: Error creating Librato space chart Foo Bar: POST https://metrics-api.librato.com/v1/spaces/158674/charts: 400  Parameter errors: streams: no metric 'cpu' exists,.
> === RUN   TestAccLibratoSpaceChart_Updated
> --- PASS: TestAccLibratoSpaceChart_Updated (6.17s)
> === RUN   TestAccLibratoSpace_Basic
> --- PASS: TestAccLibratoSpace_Basic (3.38s)
> FAIL
> exit status 1
> FAIL    github.com/hashicorp/terraform/builtin/providers/librato        19.059s
> ―
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

stack72 commented:
Hi @henrikhodne 

Have you been able to look at the comments left by @phinze? I'd love to be able to get this merged in 0.6.15

P.

jen20 commented:
Looks like a minor modification to the `Chart_Full` test to use one of the metrics created by the agent will fix the failing test here! I've rebased onto `master` and will merge this.

```
$ make testacc TEST=./builtin/providers/librato
==> Checking that code complies with gofmt requirements...
/Users/James/Code/go/bin/stringer
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/librato -v  -timeout 120m
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccLibratoSpaceChart_Basic
--- PASS: TestAccLibratoSpaceChart_Basic (6.26s)
=== RUN   TestAccLibratoSpaceChart_Full
--- PASS: TestAccLibratoSpaceChart_Full (7.30s)
=== RUN   TestAccLibratoSpaceChart_Updated
--- PASS: TestAccLibratoSpaceChart_Updated (6.72s)
=== RUN   TestAccLibratoSpace_Basic
--- PASS: TestAccLibratoSpace_Basic (3.61s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/librato	23.909s
```

jen20 commented:
Landed as `8f07a2d` - thanks @henrikhodne and all involved!

josephruscio commented:
👏

henrikhodne commented:
Thank you so much @jen20! Sorry I didn't get around to fixing this earlier.
